Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+            }
+
+            resource._url = uri.resolve(new Uri(getAbsoluteUri(this._url))).toString();
+        }
+
+        if (defined(options.queryParameters)) {
+            resource._queryParameters = combine(options.queryParameters, resource._queryParameters);
+        }
+        if (defined(options.templateValues)) {
+            resource._templateValues = combine(encodeValues(options.templateValues), resource.templateValues);
+        }
+        if (defined(options.headers)) {
+            resource.headers = combine(options.headers, resource.headers);
+        }
+        if (defined(options.responseType)) {
+            resource.responseType = options.responseType;
+        }
+        if (defined(options.method)) {
+            resource.method = options.method;
+        }
+        if (defined(options.data)) {
+            resource.data = options.data;
+        }
+        if (defined(options.overrideMimeType)) {
+            resource.overrideMimeType = options.overrideMimeType;
+        }
+        if (defined(options.proxy)) {
+            resource.proxy = options.proxy;
+        }
+        if (defined(options.request)) {
+            resource.request = options.request;
+        }
+        if (defined(options.isDirectory)) {
+            resource._isDirectory = options.isDirectory;
+        }
+        if (defined(options.retryCallback)) {
+            resource.retryCallback = options.retryCallback;
+        }
+        if (defined(options.retryAttempts)) {
+            resource.retryAttempts = options.retryAttempts;
+        }
+
+        return resource;
+    };
+
+    /**
+     * Called when a resource fails to load. This will call the retryCallback function if defined until retryAttempts is reached.
+     *
+     * @param {Error} [error] The error that was encountered.
+     *
+     * @returns {Promise<Boolean>} A promise to a boolean, that if true will cause the resource request to be retried.
+     */
+    Resource.prototype.retryOnError = function(error) {
+        var retryCallback = this.retryCallback;
+        if ((typeof retryCallback !== 'function') || (this._retryCount >= this.retryAttempts)) {
+            return when(false);
+        }
+
+        var that = this;
+        return when(retryCallback(this, error))
+            .then(function(result) {
+                ++that._retryCount;
+
+                return result;
+            });
+    };
+
+    /**
+     * Duplicates a Resource instance.
+     *
+     * @param {Resource} [result] The object onto which to store the result.
+     *
+     * @returns {Resource} The modified result parameter or a new Resource instance if one was not provided.
+     */
+    Resource.prototype.clone = function(result) {
+        if (!defined(result)) {
+            result = new Resource({
+                url : this._url
+            });
+        }
+
+        result._url = this._url;
+        result._queryParameters = this._queryParameters;
```

Comment:
Should query/templates and other non-basic types get cloned here as well?  We seem to be inconsistent throughout Cesium in this regard.  CC @shunter